hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/qjournal/client/SpyQJournalUtil.java

```diff
@@ -0,0 +1,107 @@
+import org.apache.hadoop.thirdparty.com.google.common.util.concurrent.MoreExecutors;
+import org.mockito.Mockito;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import java.util.concurrent.Executors;
+
+import static org.junit.Assert.fail;
+import static org.mockito.ArgumentMatchers.eq;
+
+/**
+ * One Util class to mock QJuournals for some UTs not in this package.
+ */
+public final class SpyQJournalUtil {
+
+  private SpyQJournalUtil() {
+  }
+
+  /**
+   * Mock a QuorumJournalManager with input uri, nsInfo and namServiceId.
+   * @param conf input configuration.
+   * @param uri input uri.
+   * @param nsInfo input nameservice info.
+   * @param nameServiceId input nameservice Id.
+   * @return one mocked QuorumJournalManager.
+   * @throws IOException throw IOException.
+   */
+  public static QuorumJournalManager createSpyingQJM(Configuration conf,
+      URI uri, NamespaceInfo nsInfo, String nameServiceId) throws IOException {
+    AsyncLogger.Factory spyFactory = (conf1, nsInfo1, journalId1, nameServiceId1, addr1) -> {
+      AsyncLogger logger = new IPCLoggerChannel(conf1, nsInfo1, journalId1, nameServiceId1, addr1);
+      return Mockito.spy(logger);
+    };
+    return new QuorumJournalManager(conf, uri, nsInfo, nameServiceId, spyFactory);
+  }
+
+  /**
+   * Try to mock one abnormal JournalNode with one empty response
+   * for getJournaledEdits rpc with startTxid.
+   * @param manager QuorumJournalmanager.
+   * @param startTxid input StartTxid.
+   */
+  public static void mockOneJNReturnEmptyResponse(
+      QuorumJournalManager manager, long startTxid, int journalIndex) {
+    List<AsyncLogger> spies = manager.getLoggerSetForTests().getLoggersForTests();
+
+    // Mock JN0 return an empty response.
+    GetJournaledEditsResponseProto responseProto = GetJournaledEditsResponseProto
+        .newBuilder().setTxnCount(journalIndex).build();
+    ListenableFuture<GetJournaledEditsResponseProto> ret = Futures.immediateFuture(responseProto);
+    Mockito.doReturn(ret).when(spies.get(journalIndex))
+        .getJournaledEdits(eq(startTxid), eq(QuorumJournalManager.QJM_RPC_MAX_TXNS_DEFAULT));
+  }
+
+  /**
+   * Try to mock one abnormal JournalNode with slow response for
+   * getJournaledEdits rpc with startTxid.
+   * @param manager input QuormJournalManager.
+   * @param startTxid input start txid.
+   * @param sleepTime sleep time.
+   * @param journalIndex the journal index need to be mocked.
+   */
+  public static void mockOneJNWithSlowResponse(
+      QuorumJournalManager manager, long startTxid, int sleepTime, int journalIndex) {
+    List<AsyncLogger> spies = manager.getLoggerSetForTests().getLoggersForTests();
+
+    ListeningExecutorService service = MoreExecutors.listeningDecorator(
+        Executors.newSingleThreadExecutor());
+    Mockito.doAnswer(invocation -> service.submit(() -> {
+      Thread.sleep(sleepTime);
```

Comment:
Same [comment as on HDFS-16659](https://github.com/apache/hadoop/pull/4560#discussion_r955289672), we shouldn't rely on timing, better to use proper synchronization